pkg/tsdb/cloudwatch/cloudwatch.go

```diff
@@ -50,26 +66,22 @@ const logStreamIdentifierInternal = "__logstream__grafana_internal__"
 var plog = log.New("tsdb.cloudwatch")
 var aliasFormat = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
 
-func init() {
-	tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
-		return newExecutor(), nil
-	})
-}
+const defaultConcurrentQueries = 4
 
 func newExecutor() *cloudWatchExecutor {
 	return &cloudWatchExecutor{
 		logsClientsByRegion: map[string]cloudwatchlogsiface.CloudWatchLogsAPI{},
+		queuesByRegion:      map[string]chan bool{},
 	}
 }
 
-// cloudWatchExecutor executes CloudWatch requests.
-type cloudWatchExecutor struct {
-	*models.DataSource
+func init() {
+	globalExecutor := newExecutor()
+	tsdb.RegisterTsdbQueryEndpoint("cloudwatch", func(ds *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
+		return globalExecutor, nil
+	})
 
-	ec2Client           ec2iface.EC2API
-	rgtaClient          resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
-	logsClientsByRegion map[string]cloudwatchlogsiface.CloudWatchLogsAPI
-	mtx                 sync.Mutex
+	live.RegisterHandler("CloudWatch", &LogQueryRunnerSupplier{})
```

Comment:
I don't really understand the workflow here -- this looks like when cloudwatch starts, it sets up an endpoint with live.  I would expect something more like

When live gets a request to: `/ds/22/path`, live will look up datasource 22, find the plugin that owns it and ask it for the config

I'm not sure the best way to have plugins register their support, for cloudwatch it does not really matter, but we soon want to do this for gRPC based plugins also

Reply:
There is a map `dsHandlerProviders`, where the datasourceID (in this case, "CloudWatch") and the corresponding `ChannelHandlerProvider` are stored when `RegisterHandler` is called. Later when`initChannel` is called, the `Namespace` parameter of `id` is used as a key for `dsHandlerProviders` to retrieve the handler provider, after which `GetHandlerForPath` is called.